Fail reading from an index until a full scan is completed

smoczy123 opened the pull request:
Add failing the read from an index until a full scan is completed.

Reference: VECTOR-43

knowack1 commented on crates/vector-store/src/db_index.rs:
```diff
@@ -61,6 +66,14 @@ impl DbIndexExt for mpsc::Sender<DbIndex> {
         rx.await.expect("internal actor should send response")
     }
+
+    async fn full_scan_progress(&self) -> u64 {
```
Consider to return percentage value

smoczy123 replied:
We would also like to provide feedback to the user about the length of the fullscan, see https://scylladb.atlassian.net/browse/VECTOR-43

smoczy123 commented on crates/vector-store/src/httproutes.rs:
```diff
@@ -309,6 +309,17 @@ async fn post_index_ann(
 
+    let scan_length = db_index.full_scan_progress().await;
+
+    if scan_length != u64::MAX {
```
DbIndex is responsible for performing the full scan, so I'm not sure if moving it to the ann() impl would work

smoczy123 replied:
And to provide proper integration tests, integration with Scylla is required, we are not there yet

knowack1 replied:
> And to provide proper integration tests, integration with Scylla is required, we are not there yet

I mean rust integration test: crates/vector-store/tests/integration. i believe there should be test that verifies the 500 response when scan is in progress.

smoczy123 commented on crates/vector-store/src/db_index.rs:
```diff
@@ -224,12 +247,18 @@ impl Statements {
                             _ = tx.send(embedding).await;
                         })
                         .await;
+                    //Safety: end > begin, and the range fits into u64
+                    scan_length.fetch_add(
+                        (end.value() - begin.value() + 1).try_into().unwrap(),
+                        std::sync::atomic::Ordering::SeqCst,
+                    );
```
Yes, but we also want to know the percentage of completed fullscan

QuerthDP reviewed (commented):
Please address the comments so I could understand the logic behind this changes.

smoczy123 commented on crates/vector-store/src/httproutes.rs:
```diff
@@ -309,6 +309,17 @@ async fn post_index_ann(
+    if scan_length != u64::MAX {
+        let msg = format!(
+            "Full scan is in progress, percentage: {:.2}%",
+            (scan_length / u64::MAX) as f64 * 100.0
```
Full length of the token range is u64::MAX

smoczy123 commented on crates/vector-store/src/db_index.rs:
```diff
@@ -185,6 +207,7 @@ impl Statements {
                 .context("range_scan_query")?,
 
             session,
+            completed_scan_length: Arc::new(AtomicU64::new(0)),
```
https://github.com/scylladb/vector-store/blob/63458bb06347c1ffaf912be31c300e828cb95ff0/crates/vector-store/src/db_index.rs#L258

QuerthDP reviewed (dismissed):
Looks good now. 
Thanks for the explanation.
Making it return the percentage made it a bit more clear as well tbh.

knowack1 commented on crates/vector-store/src/db_index.rs:
```diff
@@ -61,6 +66,14 @@ impl DbIndexExt for mpsc::Sender<DbIndex> {
         rx.await.expect("internal actor should send response")
     }
+
+    async fn full_scan_progress(&self) -> u64 {
```
return u32 (or u8) as value range is 0-100.

smoczy123 commented on crates/vector-store/src/httproutes.rs:
```diff
@@ -309,6 +309,14 @@ async fn post_index_ann(
         return (StatusCode::NOT_FOUND, "").into_response();
     };
 
+    let scan_length = db_index.full_scan_progress().await;
+
+    if scan_length != 100 {
+        let msg = format!("Full scan is in progress, percentage: {scan_length}%");
+        debug!("post_index_ann: {msg}");
+        return (StatusCode::INTERNAL_SERVER_ERROR, msg).into_response();
+    }
+
```
It's always rounded down, so it won't be an issue

ewienik replied:
I'd propose a new types `Progress` and `Percentage` which could provide statically checked values.

ewienik commented on crates/vector-store/src/db_index.rs:
```diff
@@ -224,12 +250,18 @@ impl Statements {
                         .await;
+                    //Safety: end > begin, and the range fits into u64
+                    scan_length.fetch_add(
+                        (end.value() - begin.value() + 1).try_into().unwrap(),
```
Wrapper around driver could be really hard, so providing unit test for this could be hard - db_index is an interface to the database, so it should be tested in integration tests with actual scylladb.

smoczy123 commented:
Changelog: 

- changed the type of the progress counter to u8
- moved the scan length out of the statements struct
- refactored the duplicate code in tests into new function

knowack1 commented on crates/vector-store/src/db_index.rs:
```diff
@@ -47,10 +48,32 @@ pub enum DbIndex {
+    },
+}
+
+type Progress = u64;
+
+#[derive(Clone)]
+pub struct Percentage {
+    pub value: Progress,
+    pub max: Progress,
+}
+
+impl Percentage {
+    pub fn is_done(&self) -> bool {
+        self.value == self.max
+    }
+
+    pub fn get_percentage(&self) -> u8 {
+        ((self.value as f64 / self.max as f64) * 100.0).floor() as u8
+    }
 }
```
This looks deferentially more Rusty :) Deferentially  Progress -> Percentage dependency is better than  Percentage -> Progress. Also is_done/Done suits more to Progress than Percentage.

knowack1 commented on crates/vector-store/src/db_index.rs:
```diff
@@ -47,10 +48,32 @@ pub enum DbIndex {
+
+#[derive(Clone)]
+pub struct Percentage {
+    pub value: Progress,
```
I would make props private and make ::new function

smoczy123 commented:
Changelog:

- Edited the Progress and Percentage structs according to  @ewienik proposal
- Changed the response code of the error to 503

smoczy123 commented on crates/vector-store/src/db_index.rs:
```diff
@@ -224,12 +268,18 @@ impl Statements {
                         .await;
+                    //Safety: end > begin, and the range fits into u64
+                    scan_length.fetch_add(
+                        (end.value() - begin.value() + 1).try_into().unwrap(),
```
It seems like there's no need as .abs_diff() exists and does exactly what we need
